src/trails.erl

```diff
@@ -172,13 +217,32 @@ all() ->
   end.
 
 %% @doc Fetch the trail that matches with the given path.
--spec retrieve(route_match()) -> trail() | notfound.
-retrieve(Path) ->
+-spec retrieve(PathMatch::string()) -> trail() | notfound.
+retrieve(PathMatch) ->
+  retrieve('_', PathMatch).
+
+%% @doc Fetch the trail that matches with the given host && path.
+-spec retrieve(HostMatch::route_match(),
+               PathMatch::string()) -> trail() | notfound.
+retrieve(HostMatch, PathMatch) ->
+  retrieve('_', HostMatch, PathMatch).
+
+%% @doc Fetch the trail that matches with the given server && host && path.
+-spec retrieve(Server::ranch:ref(),
+               HostMatch::route_match(),
+               PathMatch::string()) -> trail() | notfound.
+retrieve(Server, HostMatch, PathMatch) ->
   case application:get_application(trails) of
     {ok, trails} ->
-      case ets:lookup(trails, Path) of
-        [{_, Val}] -> remove_id(Val);
-        []         -> notfound
+      Key = {Server, HostMatch, PathMatch},
+      case ets:select(trails, [{{Key, '$1'}, [], ['$$']}]) of
+        % No elements found
+        [] -> notfound;
+        % One element found
+        [[Trail = #{path_match := PathMatch}]] -> remove_id(Trail);
+        % More than one element found
+        Trails when length(Trails) > 1 ->
```

Comment:
As stated above, `length` should not be used in this case. A better way to get the same effect would be…
```erlang
[_,_|_] -> throw(multiple_trails)
```